### PR TITLE
winui3 option for app project template (and update C++ template)

### PR DESCRIPTION
**Pull request by asklar**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4799)

### CODE REVIEW

File: vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj [hunk 1]
```diff
@@ -70,6 +70,7 @@
     <WholeProgramOptimization>true</WholeProgramOptimization>
     <LinkIncremental>false</LinkIncremental>
   </PropertyGroup>
+  <%=XamlProps%>
```
Comment (NickGerleman) [inline vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj:73]: I don't love that we're adding more props here.

This feels like the time to make the breaking change of pulling more props/targets into the RNW package. We do that already for bundling IIRC, but with current structure we will keep on adding changes that users will need to manually update.

E.g. Something like a Microsoft.ReactNative.targets inside of the package where we can freely add to without changing the template. (Not sure if that name is in use right now).

File: vnext/local-cli/generator-windows/index.js [hunk 2]
```diff
@@ -81,6 +81,25 @@ function copyProjectTemplateAndReplace(
   const currentUser = username.sync(); // Gets the current username depending on the platform.
   const certificateThumbprint = generateCertificate(srcPath, destPath, newProjectName, currentUser);
 
+
+  const xamlNugetPkgName = options.useWinUI3 ? 'Microsoft.WinUI' : 'Microsoft.UI.Xaml';
+  const xamlNugetPkgVersion = options.useWinUI3 ? '3.0.0-alpha.200210.0' : '2.3.191129002';
+
+  const xamlProps30 = String.raw`<Import Project="..\packages\Microsoft.WinUI.3.0.0-alpha.200210.0\build\native\Microsoft.WinUI.props" Condition="Exists('..\packages\Microsoft.WinUI.3.0.0-alpha.200210.0\build\native\Microsoft.WinUI.props')" />
```
Comment (NickGerleman) [inline vnext/local-cli/generator-windows/index.js:88]: What's the "String.raw" part for?

File: vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj (hunk 1 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj:73]: @NickGerleman this is something @jonthysell  is working on (and @dannyvv mentioned as well, and has a proposal out for).

File: vnext/local-cli/generator-windows/index.js (hunk 2 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/index.js:88]: @NickGerleman Raw string literals (e.g. don't have to escape backslashes)

File: vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj (hunk 1 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj:70]: taking this up again now that Jon's change is in. After we talked about it I think this is still the right change - the app should be in complete control of its dependencies, we should just give it a good default to start with (2.x or 3.0).

File: vnext/ReactWindows-Universal.sln [hunk 3]
```diff
@@ -16,7 +16,7 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactCommon", "ReactCommon\
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactWindowsCore", "ReactWindowsCore\ReactWindowsCore.vcxitems", "{11C084A3-A57C-4296-A679-CAC17B603145}"
 EndProject
-Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactWindowsCore", "ReactWindowsCore\ReactWindowsCore.vcxproj", "{11C084A3-A57C-4296-A679-CAC17B603144}"
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactWindowsCore-UWP", "ReactWindowsCore\ReactWindowsCore.vcxproj", "{11C084A3-A57C-4296-A679-CAC17B603144}"
```
Comment (jonthysell) [inline vnext/ReactWindows-Universal.sln:19]: Why is this renamed?

File: vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj [hunk 4]
```diff
@@ -67,6 +67,7 @@
...
```
Comment (jonthysell) [inline vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj:70]: So the expectation is that people choose WinUI at project creation?

File: vnext/local-cli/generator-windows/index.js [hunk 5]
```diff
@@ -81,6 +81,27 @@ function copyProjectTemplateAndReplace(
   const currentUser = username.sync(); // Gets the current username depending on the platform.
   const certificateThumbprint = generateCertificate(srcPath, destPath, newProjectName, currentUser);
 
+
+  const xamlNugetPkgName = options.useWinUI3 ? 'Microsoft.WinUI' : 'Microsoft.UI.Xaml';
```
Comment (jonthysell) [inline vnext/local-cli/generator-windows/index.js:85]: Are we really changing the root namespace?

File: vnext/ReactWindows-Universal.sln (hunk 3 above)
Comment (asklar) [reply vnext/ReactWindows-Universal.sln:19]: otherwise VS complains about two projects with the same name (see line 17) since now we have a vcxitems and a vcxproj

File: vnext/local-cli/generator-windows/index.js (hunk 5 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/index.js:85]: @jonthysell this line is for the name of the nuget package. Since the WinUI 3 package includes COMP in addition to XAML, the new package name is WinUI instead of MUX

File: vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj (hunk 4 above)
Comment (asklar) [reply vnext/local-cli/generator-windows/templates/cpp/proj/MyApp.vcxproj:70]: @jonthysell - Mixing winui 2.x and 3.0 isn't really supported (read: tested) so using RNW with winui2.x and then want to use winui 3 in your app or vice versa, you end up with undefined behavior AFAICT. Plus you don't want to load both WUX and MUX dlls in your process.
I could have written this in a way that exposed WinUI 3 as a build configuration option but this seemed less prone for folks to get into a bad mix situation, it's either 2.x or 3.
